quickwit-search/src/root.rs

```diff
@@ -0,0 +1,353 @@
+pub async fn root_search(
+    search_request: &SearchRequest,
+    metastore: &dyn Metastore,
+    client_pool: &Arc<SearchClientPool>,
+) -> anyhow::Result<SearchResult> {
+    let start_instant = tokio::time::Instant::now();
+
+    // Create a job for leaf node search and assign the splits that the node is responsible for based on the job.
+    let split_metadata_list = list_relevant_splits(search_request, metastore).await?;
+
+    // Create a hash map of SplitMetadata with split id as a key.
+    let split_metadata_map: HashMap<String, SplitMetadata> = split_metadata_list
+        .into_iter()
+        .map(|split_metadata| (split_metadata.split_id.clone(), split_metadata))
+        .collect();
+
+    // Create a job for fetching docs and assign the splits that the node is responsible for based on the job.
+    let leaf_search_jobs: Vec<Job> = split_metadata_map
+        .keys()
+        .map(|split_id| {
+            // TODO: Change to a better way that does not use unwrap().
+            let split_metadata = split_metadata_map.get(split_id).unwrap();
+            Job {
+                split: split_id.clone(),
+                cost: compute_split_cost(split_metadata),
+            }
+        })
+        .collect();
+
+    let assigned_leaf_search_jobs = client_pool.assign_jobs(leaf_search_jobs).await?;
+
+    let clients = client_pool.clients.read().await;
+
+    // Create search request with start offset is 0 that used by leaf node search.
+    let mut search_request_with_offset_0 = search_request.clone();
+    search_request_with_offset_0.start_offset = 0;
+    search_request_with_offset_0.max_hits += search_request.start_offset;
+
+    // Perform the query phese.
+    let mut leaf_search_handles: Vec<JoinHandle<anyhow::Result<LeafSearchResult>>> = Vec::new();
+    for (addr, jobs) in assigned_leaf_search_jobs {
+        let mut search_client = if let Some(client) = clients.get(&addr) {
+            client.clone()
+        } else {
+            return Err(anyhow::anyhow!(
+                "There is no search client for {} in the search client pool.",
+                addr
+            ));
+        };
+
+        let leaf_search_request = LeafSearchRequest {
+            search_request: Some(search_request_with_offset_0.clone()),
+            split_uris: jobs.iter().map(|job| job.split.clone()).collect(),
+        };
+
+        let handle = tokio::spawn(async move {
+            match search_client.leaf_search(leaf_search_request).await {
+                Ok(resp) => Ok(resp.into_inner()),
+                Err(err) => Err(anyhow::anyhow!(
+                    "Failed to search a leaf node due to {:?}",
+                    err
+                )),
+            }
+        });
+        leaf_search_handles.push(handle);
+    }
+    let leaf_search_responses = futures::future::try_join_all(leaf_search_handles).await?;
+
+    // Find the sum of the number of hits and merge multiple partial hits into a single partial hits.
```

Comment:
Can this lines be replaced by QuickwitCollector::merge_fruits?

Reply:
Replaced.
Please see https://github.com/quickwit-inc/quickwit/pull/163/commits/3c6c30957b6698233ff41c60797f8bcd0b31dae5.

Reply:
no need to mention the specific commit. This is a lot of work for you.